src/charm.py

```diff
@@ -121,92 +126,116 @@ def _get_database_relation_info(self) -> dict:
+            logger.info("Hydra service is absent. Deferring database created event.")
 
+    def _on_database_changed(self, event) -> None:
+        """Event Handler for database changed event."""
         if not self._container.can_connect():
             event.defer()
             logger.info("Cannot connect to Hydra container. Deferring the event.")
             self.unit.status = WaitingStatus("Waiting to connect to Hydra container")
             return
 
-        self._update_layer()
-
-        if not self.unit.is_leader():
-            # TODO: Observe leader_elected event
-            logger.info("Unit does not have leadership")
-            self.unit.status = WaitingStatus("Waiting for leadership")
-            return
-
-        self._run_sql_migration()
+        self.unit.status = MaintenanceStatus("Updating database details")
 
-        self.unit.status = ActiveStatus()
-
-    def _run_sql_migration(self) -> None:
-        """Runs a command to create SQL schemas and apply migration plans."""
-        process = self._container.exec(
-            ["hydra", "migrate", "sql", "-e", "--config", self._hydra_config_path, "--yes"],
-            timeout=20.0,
-        )
         try:
-            stdout, _ = process.wait_output()
-            logger.info(f"Executing automigration: {stdout}")
-        except ExecError as err:
-            logger.error(f"Exited with code {err.exit_code}. Stderr: {err.stderr}")
-            self.unit.status = BlockedStatus("Database migration job failed")
-
-    def _on_hydra_pebble_ready(self, event) -> None:
-        """Event Handler for pebble ready event."""
-        self._update_container(event)
-
-    def _on_db_events(self, db_event) -> None:
-        """Event Handler for database-related events."""
-        logger.info("Retrieved database details")
-        self._update_container(db_event)
+            self._container.get_service(self._container_name)
+            logger.info("Updating Hydra config and restarting service")
+            self._container.push(self._hydra_config_path, self._config, make_dirs=True)
+            self._container.restart(self._container_name)
+            self.unit.status = ActiveStatus()
+        except (ModelError, RuntimeError):
+            event.defer()
+            self.unit.status = WaitingStatus("Waiting for Hydra service")
+            logger.info("Hydra service is absent. Deferring database created event.")
+
+    def _on_run_migration(self, event: ActionEvent) -> None:
+        """Runs the migration as an action response."""
+        logger.info("Executing database migration initiated by user")
+        self._run_sql_migration(set_timeout=False)
+
+    def _on_database_relation_departed(self, event) -> None:
+        """Event Handler for database relation departed event."""
+        logger.error("Missing required relation with postgresql")
+        self.model.unit.status = BlockedStatus("Missing required relation with postgresql")
```

Comment:
Should we update the Hydra config here (ie remove the `dsn` or stop the service). I assume that a departed relation means that the database will be gone and that Hydra will be trying to call the database APIs and get errors.